Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,547 @@
+                    engineEnablements[engine] = enabled
+                }
+            }
+        }
+
+        if !engineEnablements.isEmpty {
+            let enabled = engineEnablements.compactMap { $0.value ? $0.key : nil }
+            logger.log("engines to enable: \(enabled)",
+                       level: .info,
+                       category: .sync)
+
+            let disabled = engineEnablements.compactMap { !$0.value ? $0.key : nil }
+            let msg = "engines to disable: \(disabled)"
+            logger.log(msg,
+                       level: .info,
+                       category: .sync)
+        }
+        return engineEnablements
+    }
+
+    public class ScopedKeyError: MaybeErrorType {
+        public let description = "No key data found for scope."
+    }
+
+    public class EncryptionKeyError: MaybeErrorType {
+        public let description = "Failed to get stored key."
+    }
+
+    public class DeviceIdError: MaybeErrorType {
+        public let description = "Failed to get deviceId."
+    }
+
+    public class NoTokenServerURLError: MaybeErrorType {
+        public let description = "Failed to get token server endpoint url."
+    }
+
+    public class EngineAndKeyRetrievalError: MaybeErrorType {
+        public let description = "Failed to get sync engine and key data."
+    }
+
+    private func getEnginesAndKeys(engines: [String]) -> Deferred<Maybe<([EngineIdentifier],
+                                                                         [String: String])>> {
+        let deferred = Deferred<Maybe<([EngineIdentifier], [String: String])>>()
+        var localEncryptionKeys: [String: String] = [:]
+        var rustEngines: [String] = []
+        var registeredPlaces: Bool = false
+
+        for engine in engines {
+            switch engine {
+            case "tabs":
+                profile?.tabs.registerWithSyncManager()
+                rustEngines.append(engine)
+            case "passwords":
+                profile?.logins.registerWithSyncManager()
+                if let key = try? profile?.logins.getStoredKey() {
+                    localEncryptionKeys[engine] = key
+                    rustEngines.append(engine)
+                } else {
+                    logger.log("Login encryption key could not be retrieved for syncing",
+                               level: .warning,
+                               category: .sync)
+                }
+            case "bookmarks", "history":
+                if !registeredPlaces {
+                    profile?.places.registerWithSyncManager()
+                    registeredPlaces = true
+                }
+                rustEngines.append(engine)
+            default:
+                continue
+            }
+        }
+
+        deferred.fill(Maybe(success: (rustEngines, localEncryptionKeys)))
+        return deferred
+    }
+
+    private func syncRustEngines(why: MozillaAppServices.SyncReason,
```

Comment:
There's an argument that this function is too massive. How testable is it? Can it be split up into smaller functions that are more easily parsed and that simply following the logic?

Reply:
I'm planning to break it into smaller pieces when I address #13543